### ISSUE TITLE
Rule suggestion - Prevent using built-in prototype properties on an object

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#2693]
Prevent use of built-in prototype properties on an object as these won't work on objects without a prototype, eg. object created with `Object.create(null)`. You never know when you'll get an object like that, so better be safe than sorry. This rule would have prevented #2688.

Should prevent the use of these:

```js
Object.getOwnPropertyNames(Object.prototype);
//=> ["constructor", "toString", "toLocaleString", "valueOf", "hasOwnProperty", "isPrototypeOf", "propertyIsEnumerable", "__defineGetter__", "__lookupGetter__", "__defineSetter__", "__lookupSetter__", "__proto__"]
```

Example of misuse:

```js
var obj = {};
obj.foo = true;
obj.hasOwnProperty('foo');
//=> true
```

Which will fail here:

```js
var obj = Object.create(null);
obj.foo = true;
obj.hasOwnProperty('foo');
//=> 'Uncaught TypeError: obj.hasOwnProperty is not a function'
```

Instead this should be used:

```js
var obj = Object.create(null);
obj.foo = true;
Object.prototype.hasOwnProperty.call(obj, 'foo');
//=> 'foo'
```

**Comment by nzakas** [comment]
You're suggesting to warn 100% of the time when these methods are not called from `Object.prototype`?

**Comment by sindresorhus** [comment]
Yes

**Comment by ilyavolodin** [comment]
I think it's a nice rule to have, but this should definitely be off by default (not that it matters after 1.0.0). I wonder if this could be made smarter though. If, for example, escope, tracked how the variable was created, this could flag only use of prototype properties on objects created through `Object.create`. But I'm pretty sure escope doesn't track that.

**Comment by gyandeeps** [comment]
FYI, if its created like this `Object.create(null)`, then only it applies.

**Comment by sindresorhus** [comment]
> this could flag only use of prototype properties on objects created through Object.create

There's no way to track that reliably as the object might come from another dependency or dynamically created at runtime.

**Comment by ilyavolodin** [comment]
You are right, but I can dream, can't I?:-) But I still think it's a good rule to have.

**Comment by michaelficarra** [comment]
I always pull Array and Object prototype members off fresh instances: `[].push.apply(...)` and `{}.hasOwnProperty.call(...)`. I suggest both patterns are allowed.

**Comment by sindresorhus** [comment]
> I always pull Array and Object prototype members off fresh instances: [].push.apply(...) and {}.hasOwnProperty.call(...). I suggest both patterns are allowed.

:+1: I do that too. `{}.hasOwnProperty.call(...)` => `({}).hasOwnProperty.call(...)`. OT, but V8 optimizes this, so it has the same perf as using the prototype directly.

**Comment by nzakas** [comment]
Okay, I think the easiest thing to check for is that some list of methods can only be called using `call` or `apply`, otherwise we need to worry about the various objects on which the methods reside. Would that work?

**Comment by michaelficarra** [comment]
Can we have a concrete proposal for the list of names? I'm fine with catching `hasOwnProperty`, but `toString`, `push`, `valueOf`, etc. are probably not a good idea. It may need to be configurable.

**Comment by denis-sokolov** [comment]
Correct me if I'm wrong, the use case we're covering is when `Object.create(null)` is used as a simple replacement for a `Map`. In that context, clearly `hasOwnProperty` and `propertyIsEnumerable` can be used, but `toString`, `push`, `valueOf` and `isPrototypeOf` are unlikely to be used on it.
Thus, perhaps covering `hasOwnProperty` and `propertyIsEnumberable` is enough?

**Comment by nzakas** [comment]
@sindresorhus provided his desired list in the original description.

**Comment by michaelficarra** [comment]
I didn't think that was a serious suggestion. `toString` is too commonly overridden. Also, the `__` accessors can't be simply `call`ed or `apply`ed. `Object.getOwnPropertyDescriptor(Object.prototype, "__proto__").get.call(obj)` is *not* a common pattern.

**Comment by nzakas** [comment]
@sindresorhus was your suggestion serious? :)

**Comment by sindresorhus** [comment]
It was just to be completely safe, but I guess these are enough:

- hasOwnProperty
- isPrototypeOf
- propertyIsEnumerable

**Comment by mysticatea** [comment]
To think the name of this rule is difficult to me...

**Comment by ilyavolodin** [comment]
Something like `no-prototype-properties`?

**Comment by nzakas** [comment]
Hmm, not quite right. It's not being prototype properties.

**Comment by nzakas** [comment]
Sorry, not banning prototype properties.

**Comment by gyandeeps** [comment]
@eslint/eslint-team we left this half way hanging....

**Comment by nzakas** [comment]
Still looking for naming suggestions.

**Comment by sindresorhus** [comment]
Some name suggestions:

- no-builtin-proto-props
- no-unsafe-proto-props

**Comment by alberto** [comment]
Does this still fit our [criteria for new rules](http://eslint.org/docs/developer-guide/contributing/new-rules#core-rule-guidelines)? 
Do we have a champion? 
@sindresorhus are you willing to implement this yourself?

**Comment by sindresorhus** [comment]
I'm willing to do the work if this is still wanted and we can settle on a name.

**Comment by nzakas** [comment]
Looks like we need a champion @eslint/eslint-team

**Comment by DrewML** [comment]
I'm willing to help with this one. I have a quick POC for it [here](http://astexplorer.net/#/a2NFDuLsDy/1) (assuming I understood the current requirements correctly).

Do we care about the (very unlikely) scenario that someone uses this pattern?

```javascript
var foo = {};
foo.bar = true;
var hasOwnProperty = foo.hasOwnProperty;
hasOwnProperty.call(foo, 'bar');
```

I'll go through the contributor docs (hopefully) tomorrow and try to get a PR in with tests.

**Comment by mysticatea** [comment]
I'm a champion.

`hasOwnProperty` is often used with `Object.create(null)` in order to use an object as a key-value map.
Map exists in ES2015, but `Object.create(null)` is still used in many places. For example, `querystring` module of Node: https://github.com/nodejs/node/wiki/Breaking-changes-between-v5-and-v6#querystring

@DrewML Awesome! Looks good to me.

**Comment by BYK** [comment]
I know I'm late to the game but I don't think this really belongs to the core since it has a quite narrow use case from what I see. My 2 cents is to make this as a plugin.

**Comment by DrewML** [comment]
@mysticatea @sindresorhus Any thoughts re: whether this needs to be in core vs. a plugin?

**Comment by mysticatea** [comment]
I think using `Object.create(null)` is one of well-known practices in ES5. In Node, `querystring.parse()` returns an object which is created by `Object.create(null)`. In ESLint, several rules are using `Object.create(null)`. So this rule would help to avoid `TypeError`s.

**Comment by nzakas** [comment]
Sounds like we have some disagreement over whether this belongs in the core or not. Per our guidelines, a dissent needs to be resolved before we can implement. 

Remember: the bar on whether a rule should be included in the core is now quite high. We need to consider the rule to be "extremely important" to include in the core. @mysticatea do you feel like this rule is extremely important to include?

**Comment by nzakas** [comment]
@eslint/eslint-team we need some more opinions here.

**Comment by ilyavolodin** [comment]
I'll support this. It's a pretty important rule rule that's hard for new users to understand without learning more about prototypical inheritance. So I think it can save people from creating a lot of errors.

**Comment by pedrottimark** [comment]
For your info, there is a somewhat related accepted rule in progress https://github.com/eslint/eslint/pull/5872 for https://github.com/eslint/eslint/issues/3218

If I understand the options correctly, the restrictive depends on names of variables which represent instances. That seems impractical for application code, but might work well in micro libraries if there is a limited amount of code with strong naming convention.

**Comment by DrewML** [comment]
> If I understand the options correctly, the restrictive depends on names of variables which represent instances.

That's the impression I get as well. If that is the case, I think this rule still has merit.

**Comment by platinumazure** [comment]
I'd love to see this in core since `Object.create(null)` is standard and people are starting to use it more. It's completely opt-in (like everything else in ESLint), so no worry about disrupting users.

**Comment by mysticatea** [comment]
I think this is pretty important because inheriting from `null` is special thing in JavaScript -- I guess, people that are familiar with other language but not familiar with JavaScript, are hard to guess `obj.hasOwnProperty` might not exist. And some standard library (e.g. `querystring`) is using `Object.create(null)`.

**Comment by nzakas** [comment]
OK 👍

**Comment by DrewML** [comment]
Cool! Will start working on the PR for this tonight.

**Comment by mysticatea** [comment]
Thank you everyone!

**Comment by DrewML** [comment]
PR submitted :)

**Comment by ljharb** [comment]
The commit message here doesn't mention the rule name, so its' not in the changelog :-/ i think it's "no-prototype-builtins"?

**Comment by mysticatea** [comment]
Yes: http://eslint.org/blog/2016/05/eslint-v2.11.0-released#new-rules
